app-backend/db/src/test/scala/com/azavea/rf/database/ProjectDaoSpec.scala

```diff
@@ -274,4 +274,197 @@ class ProjectDaoSpec extends FunSuite with Matchers with Checkers with DBTestCon
+        ) => {
+          val projectPermissionsIO = for {
+            projMiscInsert <- fixUpProjMiscInsert(userTeamOrgPlat, project)
+            (projectInsert, dbUserTeamOrgPlat) = projMiscInsert
+            dbGrantedUser <- UserDao.create(grantedUser)
+            dbUserGrantedTeamOrgPlat = dbUserTeamOrgPlat.copy(_1 = dbGrantedUser)
+            acrInsert1 = fixUpObjectAcr(acr1, dbUserGrantedTeamOrgPlat)
+            acrInsert2 = fixUpObjectAcr(acr2, dbUserGrantedTeamOrgPlat)
+            acrInsert3 = fixUpObjectAcr(acr3, dbUserGrantedTeamOrgPlat)
+            permissions <- ProjectDao.addPermissionsMany(projectInsert.id, List(
+              acrInsert1,
+              acrInsert2,
+              acrInsert3
+            ))
+          } yield { (permissions, List(acrInsert1, acrInsert2, acrInsert3).map(Some(_))) }
+
+          val (permissions, acrOptionList) = projectPermissionsIO.transact(xa).unsafeRunSync
+
+          assert(permissions.diff(acrOptionList).length == 0,
+            "Inserting a list of permissions to a project should get back the same list of them")
+          true
+        }
+      }
+    }
+  }
+
+  // addPermission and replacePermissions
+  test("add a permission to a project and replace it with many others") {
+    check {
+      forAll {
+        ( userTeamOrgPlat: (User.Create, Team.Create, Organization.Create, Platform),
+          acr1: ObjectAccessControlRule,
+          acr2: ObjectAccessControlRule,
+          acr3: ObjectAccessControlRule,
+          project: Project.Create,
+          grantedUser: User.Create
+        ) => {
+          val projectReplacePermissionsIO = for {
+            projMiscInsert <- fixUpProjMiscInsert(userTeamOrgPlat, project)
+            (projectInsert, dbUserTeamOrgPlat) = projMiscInsert
+            dbGrantedUser <- UserDao.create(grantedUser)
+            dbUserGrantedTeamOrgPlat = dbUserTeamOrgPlat.copy(_1 = dbGrantedUser)
+            acrInsert1 = fixUpObjectAcr(acr1, dbUserGrantedTeamOrgPlat)
+            permAdded <- ProjectDao.addPermission(projectInsert.id, acrInsert1)
+            acrInsert2 = fixUpObjectAcr(acr2, dbUserGrantedTeamOrgPlat)
+            acrInsert3 = fixUpObjectAcr(acr3, dbUserGrantedTeamOrgPlat)
+            permReplaced <- ProjectDao.replacePermissions(projectInsert.id, List(
+              acrInsert2,
+              acrInsert3
+            ))
+          } yield { (
+            permAdded,
+            List(Some(acrInsert1)),
+            permReplaced,
+            List(acrInsert2, acrInsert3).map(Some(_))
+          ) }
+
+          val (permAdded, acrAdded, permReplaced, acrReplaced) =
+            projectReplacePermissionsIO.transact(xa).unsafeRunSync
+
+          assert(permAdded.diff(acrAdded).length == 0,
+            "Inserting a permission to a project should get the permission back")
+          assert(permReplaced.diff(acrReplaced).length == 0,
+            "Replacing project permissions should get the replaced permissions back")
+          true
```

Comment:
I think for completeness, you also need to test that `acrReplaced` is consistent with calling `listUserActions` after the replacement -- confirming that we get the ACRs back doesn't necessarily confirm that the initial ACR has been replaced